File: first-architecture/Invokers/dependency-providers/InvokerLOG/function/LOG-js-function/logger.js
const fs = require("fs");

/**
 * Logs a message with a given timestamp and severity level to a file.
 * @param {string} timestamp - The timestamp of the log (e.g., "2024-11-27T12:00:00Z").
 * @param {string} message - The log message.
 * @param {string} severity - The severity level (e.g., "info", "warn", "error").
 * @returns {boolean} - True if the logging operation was successful, false otherwise.
 */
function logMessageToFile(timestamp, message, severity) {
    const validSeverities = ["info", "warn", "error"];

    // Validate severity level
    if (!validSeverities.includes(severity)) {
        console.error("Invalid severity level. Valid options are: info, warn, error.");
        return false;
    }

    // Validate timestamp
    const date = new Date(timestamp);
    if (isNaN(date.getTime())) {
        console.error("Invalid timestamp format. Use ISO 8601 format (e.g., yyyy-mm-ddThh:mm:ssZ).");
        return false;
    }

    const logEntry = `[${timestamp}] [${severity.toUpperCase()}]: ${message}\n`;

    try {
        // Append the log entry to the logs.txt file
        fs.appendFileSync("logs.txt", logEntry, "utf8");
        return true; // Logging succeeded
    } catch (error) {
        console.error("Logging operation failed:", error.message);
        return false; // Logging failed
    }
}

// Get input from the command line
let params
try{
params = JSON.parse(process.argv[2]);
} catch (error) {
    console.error("Invalid input. Please provide the correct number of arguments.");
    process.exit(1);}
// Perform the logging operation
const result = logMessageToFile(params.timestamp, params.message, params.severity);
// Indicate success or failure
if (result) {
    console.log('logging operation succeeded');
} else {
    console.error("logging operation failed.");}

File: first-architecture/Invokers/dependency-providers/InvokerDP/go/LOG-js-function/logger.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"
	"time"
)

func logMessageToFile(timestamp, message, severity string) bool {
	validSeverities := []string{"info", "warn", "error"}

	// Validate severity level
	if !contains(validSeverities, severity) {
		fmt.Println("Invalid severity level. Valid options are: info, warn, error.")
		return false
	}

	// Validate timestamp
	_, err := time.Parse(time.RFC3339, timestamp)
	if err != nil {
		fmt.Println("Invalid timestamp format. Use ISO 8601 format (e.g., yyyy-mm-ddThh:mm:ssZ).")
		return false
	}

	logEntry := fmt.Sprintf("[%s] [%s]: %s\n", timestamp, strings.ToUpper(severity), message)

	// Append the log entry to the logs.txt file
	err = appendToFile("logs.txt", logEntry)
	if err != nil {
		fmt.Println("Logging operation failed:", err.Error())
		return false
	}
	return true // Logging succeeded
}

func contains(slice []string, item string) bool {
	for _, v := range slice {
		if v == item {
			return true
		}
	}
	return false
}

func appendToFile(filename, data string) error {
	f, err := os.OpenFile(filename, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		return err
	}
	defer f.Close()

	if _, err := f.WriteString(data); err != nil {
		return err
	}
	return nil
}

func main() {
	var params struct {
		Timestamp string `json:"timestamp"`
		Message   string `json:"message"`
		Severity  string `json:"severity"`
	}

	// Get input from the command line
	if len(os.Args) < 2 {
		fmt.Println("Invalid input. Please provide the correct number of arguments.")
		os.Exit(1)
	}

	err := json.Unmarshal([]byte(os.Args[1]), &params)
	if err != nil {
		fmt.Println("Invalid input. Please provide the correct number of arguments.")
		os.Exit(1)
	}

	// Perform the logging operation
	result := logMessageToFile(params.Timestamp, params.Message, params.Severity)
	// Indicate success or failure
	if result {
		os.Stdout.WriteString("logging operation succeeded")
	} else {
		os.Stdout.WriteString("logging operation failed.")
	}
}
